### PR TITLE
Draw tabs in max layout

### LINKED ISSUE #1299: Show stacked window titles
This idea comes from older tiled WMs like Ion and PWM. In those, when you had multiple windows stacked in the same frame, the space for the window title was segmented to display the title of each stacked window.

[Ion example](https://tuomov.iki.fi/software/ion/ion3-1.png)
[PWM example](https://tuomov.iki.fi/software/ion/pwm-3.jpg)
[Haiku OS example](https://www.haiku-os.org/docs/userguide/en/images/gui-images/gui-s+t.gif)

With herbstluftwm, you only see the top window's title, so there is no indication that there are other windows behind it. I often end up hitting the keybinding to cycle windows in a frame just to see if I left one back there.

### PR DESCRIPTION
**Pull request by t-wissmann**
This draws the hidden windows in max layout as tabs if the setting
tabbed_max is set (which is the default).

This directly closes #1299 and also implements a suggestion mentioned in
#1097 and in various other places throughout the years.

### CODE REVIEW

**Comment by t-wissmann** [comment]
This is still work in progress, because it still lacks:
  - [ ] tests
  - [ ] a proper commit message
  - [ ] mouse interaction (maybe for a separate PR)

**Comment by blarz** [comment]
That looks awesome already!

**Comment by The-Compiler** [comment]
FWIW, [from Reddit](https://www.reddit.com/r/herbstluftwm/comments/qzp86t/comment/hlr9gbw/?utm_source=share&utm_medium=web2x&context=3)

> Will this be toggle-able feature? Because I'm really used to the current one

Also, personally, I can't wait for this! :tada:

**Comment by t-wissmann** [comment]
> FWIW, [from Reddit](https://www.reddit.com/r/herbstluftwm/comments/qzp86t/comment/hlr9gbw/?utm_source=share&utm_medium=web2x&context=3)
> 
> > Will this be toggle-able feature? Because I'm really used to the current one

Alright, I've added an option to reenable spacebar heating.

By the way, the doc diff [in the comment above](https://github.com/herbstluftwm/herbstluftwm/pull/1381#issuecomment-975606981) still misses the `doc` string, because of a bug that is fixed in #1384.

> Also, personally, I can't wait for this! tada

I hope I can finish the PR today.

**Comment by blarz** [comment]
I am running hlwm with this for a couple of days now and it works great!

One thing I noticed though with the clickable tab: the clickable region seems to be a bit larger than the actual drawn tab? I noticed this when I tried to click on the "File" dialog of my Editor and suddenly the tab was changed.

**Comment by t-wissmann** [comment]
> I am running hlwm with this for a couple of days now and it works great!
> 
> One thing I noticed though with the clickable tab: the clickable region seems to be a bit larger than the actual drawn tab? I noticed this when I tried to click on the "File" dialog of my Editor and suddenly the tab was changed.

Thanks for testing it already! The issue was that I didn't check the window of the click event correctly (half of the `==`-comparison was missing...). It's fixed now and has a test case. Now I only need to write the remaining test cases.

**Comment by blarz** [comment]
> > I am running hlwm with this for a couple of days now and it works great!
> > One thing I noticed though with the clickable tab: the clickable region seems to be a bit larger than the actual drawn tab? I noticed this when I tried to click on the "File" dialog of my Editor and suddenly the tab was changed.
> 
> Thanks for testing it already! The issue was that I didn't check the window of the click event correctly (half of the `==`-comparison was missing...). It's fixed now and has a test case. Now I only need to write the remaining test cases.

Nice, this fixed it indeed, thank you!

**Comment by t-wissmann** [comment]
> Nice, this fixed it indeed, thank you!

Great! By the way, I've just added the remaining test cases, so I'd say the PR is ready :-)

File: tests/test_decorations.py [hunk 1]
```diff
@@ -254,3 +254,144 @@ def test_frame_holes_for_pseudotiled_client(hlwm, x11, frame_bg_transparent):
...
+    # split twice to make tab area smaller and screenshots faster :-)
+    hlwm.call(['split', 'bottom'])
+    hlwm.call(['split', 'bottom'])
+
+    winhandle = x11.window(running_clients[0])
+    img = x11.decoration_screenshot(winhandle)
+    color_count = img.color_count_dict()
+    assert urgent_color not in color_count
+
+    # make one of the unfocsed tabs urgent
```
Comment (blarz) [inline tests/test_decorations.py:312]: typo "unfocused"
```suggestion
    # make one of the unfocused tabs urgent
```

File: tests/test_decorations.py [hunk 2]
```diff
@@ -254,3 +254,144 @@ def test_frame_holes_for_pseudotiled_client(hlwm, x11, frame_bg_transparent):
...
+@pytest.mark.parametrize("running_clients_num", [4])
+def test_decoration_click_changes_tab(hlwm, mouse, running_clients, running_clients_num):
+    hlwm.call(['load', '(clients max:0 {})'.format(' '.join(running_clients))])
+    hlwm.attr.settings.tabbed_max = True
+    hlwm.attr.theme.title_height = 10
+
+    geo = hlwm.attr.clients.focus.decoration_geometry()
+    tabbar_top_left = geo.topleft()
+    tabbar_bottom_right = geo.topleft() + Point(geo.width, hlwm.attr.theme.title_height())
+    print(f"tl {tabbar_top_left}")
```
Comment (blarz) [inline tests/test_decorations.py:361]: Do you want to keep the print() statements in this test case?

**Review by blarz — CHANGES_REQUESTED** [review]
This looks great!

Just two minor "issues"

File: tests/test_decorations.py (hunk 1 above)
Comment (t-wissmann) [reply tests/test_decorations.py]: Fixed, thanks!

File: tests/test_decorations.py (hunk 2 above)
Comment (t-wissmann) [reply tests/test_decorations.py]: No, I forgot to remove them. now they're gone. Thanks!

File: tests/test_decorations.py (hunk 1 above)
Comment (blarz) [reply tests/test_decorations.py]: Thanks!

File: tests/test_decorations.py (hunk 2 above)
Comment (blarz) [reply tests/test_decorations.py]: Thanks! But there is still one printing the ratio, can you remove that one, too?

Comment (t-wissmann) [reply tests/test_decorations.py]: sorry; now it's gone :-)

Comment (blarz) [reply tests/test_decorations.py]: Thanks :)

**Review by blarz — APPROVED** [review]
This is awesome, thanks!

**Comment by t-wissmann** [comment]
> This is awesome, thanks!

Thanks a lot :-)

**Pull request merged**